src/lib/core/internal/internals.ts

```diff
@@ -0,0 +1,71 @@
+import { Observable } from 'rxjs';
+
+/**
+ * @property type - Dispatched event type
+ * @property payload - Paylaod to dispatch
+ */
+export interface DispatchedEvent<T = unknown> {
+    type: string;
+    payload?: T;
+}
+
+/**
+ * Methods decorated with `@Dispatch()` decorator can return plain object or `Promise` or `Observable`
+ */
+export type WrappedDispatchedEvent = Observable<DispatchedEvent> | Promise<DispatchedEvent> | DispatchedEvent;
+
+/**
+ * Event can be a plain object or an instance of some class
+ *
+ * @internal
+ * @param event - Dispatched event
+ * @returns - True if plain object or instance constructor has `type` property
+ */
+export function hasTypeProperty<T extends Object>(event: T): boolean {
+    return event.hasOwnProperty('type') || event.constructor.hasOwnProperty('type');
+}
+
+/**
+ * @internal
+ * @param event - Dispatched event
+ * @returns - True if event is an object and has `type` property
+ */
+export function isValidEvent<T extends Object>(event: T): boolean {
+    return !!event && typeof event === 'object' && hasTypeProperty<T>(event);
+}
+
+/**
+ * Descriptor exists only in case of method decorating
+ *
+ * @internal
+ * @param descriptor - Property descriptor
+ * @returns - True if descriptor exists
+ */
+export function descriptorExists(descriptor: TypedPropertyDescriptor<Function> | undefined): boolean {
+    return !!descriptor && descriptor.hasOwnProperty('value');
+}
+
+/**
+ * @internal
+ * @param event - Dispatched event
+ * @returns - True if event is just a plain object without constructor
+ */
+export function eventIsPlainObject<T extends Object>(event: T): boolean {
+    return event.constructor === Object;
+}
+
+/**
+ * @param target - Target to check if it's an `Observable`
+ * @returns - True if target is a stream
+ */
+export function isObservable(target: any): target is Observable<unknown> {
+    return !!target && typeof target.subscribe === 'function';
+}
+
+/**
+ * @param target - Target to check if it's a `Promise`
+ * @returns - True if target is a `Promise`
+ */
+export function isPromise(target: any): target is Promise<unknown> {
+    return !!target && typeof target.then === 'function';
```

Comment:
```ts
target instanceof Promise
```

or

```ts
Promise.resolve(target) == target
```

Reply:
@splincode I think that the existing check may be better.
Maybe the dev is not using the ES6 Promise... if we want to support any promise that implements the A+ specification for Promises then there is the concept of something that is 'thenable' (just like above). I'm not sure what else you would need to check for to test for the A+ specification...
https://promisesaplus.com/

What does this check for/how does this work?
```TS
Promise.resolve(target) == target
```
Seems like it could be expensive so if we use this approach then the result may need to be cached after first check??

Reply:
Ok, either so

```ts
return !!target && typeof target.then === 'function'
```
so

```ts
return target instanceof Promise;
```

Reply:
@splincode Agreed
![image](https://user-images.githubusercontent.com/7337691/49000931-c2611a00-f163-11e8-9924-20ef7df19456.png)